libs/types/src/pools.rs

```diff
@@ -37,3 +42,314 @@ pub enum PoolRegistrationStatus {
+	/// Account that the fees are sent to
+	pub destination: AccountId,
+
+	/// Account that can update this fee
+	pub editor: PoolFeeEditor<AccountId>,
+
+	/// Amount of fees that can be charged
+	pub fee_type: PoolFeeType<Balance, Rate>,
+}
+
+impl<AccountId, Balance, FeeId, Rate> PoolFee<AccountId, FeeId, PoolFeeAmounts<Balance, Rate>>
+where
+	Balance: Default,
+{
+	pub fn from_info(fee: PoolFeeInfo<AccountId, Balance, Rate>, fee_id: FeeId) -> Self {
+		let payable = match fee.fee_type {
+			PoolFeeType::ChargedUpTo { .. } => Some(Balance::default()),
+			PoolFeeType::Fixed { .. } => None,
+		};
+		let amount = PoolFeeAmounts {
+			payable,
+			fee_type: fee.fee_type,
+			pending: Balance::default(),
+			disbursement: Balance::default(),
+		};
+
+		Self {
+			amounts: amount,
+			destination: fee.destination,
+			editor: fee.editor,
+			id: fee_id,
+		}
+	}
+}
+
+/// The editor enum of pool fees
+#[derive(Debug, Encode, Decode, TypeInfo, MaxEncodedLen, PartialEq, Eq, Clone)]
+
+pub enum PoolFeeEditor<AccountId> {
+	Root,
+	Account(AccountId),
+}
+
+/// The fee amount wrapper type
+#[derive(Debug, Encode, Decode, TypeInfo, MaxEncodedLen, PartialEq, Eq, Clone)]
+
+pub enum PoolFeeType<Balance, Rate> {
+	/// A fixed fee is deducted automatically every epoch
+	Fixed { limit: PoolFeeAmount<Balance, Rate> },
+
+	/// A fee can be charged up to a limit, paid every epoch
+	ChargedUpTo { limit: PoolFeeAmount<Balance, Rate> },
+}
+
+/// The pending fee amount wrapper type
+#[derive(Debug, Encode, Decode, TypeInfo, MaxEncodedLen, PartialEq, Eq, Clone)]
+
+pub struct PoolFeeAmounts<Balance, Rate> {
+	pub fee_type: PoolFeeType<Balance, Rate>,
+	pub pending: Balance,
+	pub disbursement: Balance,
+	pub payable: Option<Balance>,
+}
+
+impl<Balance, Rate> PoolFeeAmounts<Balance, Rate> {
+	pub fn limit(&self) -> &PoolFeeAmount<Balance, Rate> {
+		match &self.fee_type {
+			PoolFeeType::Fixed { limit } | PoolFeeType::ChargedUpTo { limit } => limit,
+		}
+	}
+}
+
+/// The fee amount
+#[derive(Debug, Encode, Decode, TypeInfo, MaxEncodedLen, PartialEq, Eq, Clone)]
+pub enum PoolFeeAmount<Balance, Rate> {
+	ShareOfPortfolioValuation(Rate),
+	// TODO: AmountPerSecond(Balance) might be sufficient
+	AmountPerYear(Balance),
+	AmountPerMonth(Balance),
+	AmountPerSecond(Balance),
+}
+
+/// The priority segregation of pool fees
+///
+/// NOTE: Whenever a new variant is added, must bump
+/// [cfg_primitives::MAX_FEES_PER_POOL].
+#[derive(
+	Debug, Encode, Decode, EnumIter, EnumCount, TypeInfo, MaxEncodedLen, PartialEq, Eq, Clone, Copy,
+)]
+pub enum PoolFeeBucket {
+	/// Fees that are charged first, before any redemptions, investments,
+	/// repayments or originations
+	Top,
+	// Future: AfterTranche(TrancheId)
+}
+
+impl<Balance, Rate, Time> FeeAmountProration<Balance, Rate, Time> for PoolFeeAmount<Balance, Rate>
+where
+	Rate: FixedPointNumberExtension,
+	Balance: From<Seconds> + FixedPointOperand + sp_std::ops::Div<Output = Balance>,
+	Time: Into<Seconds>,
```

Comment:
Use `IntoSeconds` (located in cfg_traits) instead to make the correct arithmetic conversion. If not, it's just casting from one integer to another.

Reply:
Option 2. If you specify `Seconds` as the type used, then there is no need to be generic.

Reply:
Thanks for pointing this out! I removed the `Time` generic because the `epoch_duration` is already denominated in `Seconds`. So `into_seconds()` actually cuts of 3 digits and breaks all calculations 🫠 Fixed in 7923f3c9a0dd302b0a3e8deb76b82878adf6fa1f